internal/service/dynamodb/global_secondary_index.go

```diff
@@ -0,0 +1,934 @@
+	} else if !data.OnDemandThroughputs.IsNull() && !data.OnDemandThroughputs.IsUnknown() {
+		var odts []onDemandThroughputModel
+		response.Diagnostics.Append(data.OnDemandThroughputs.ElementsAs(ctx, &odts, false)...)
+		if len(odts) > 0 {
+			v := map[string]any{
+				"max_read_request_units":  int(odts[0].MaxReadRequestUnits.ValueInt64()),
+				"max_write_request_units": int(odts[0].MaxWriteRequestUnits.ValueInt64()),
+			}
+			action.OnDemandThroughput = expandOnDemandThroughput(v)
+		}
+	}
+
+	var wts []warmThroughputModel
+	response.Diagnostics.Append(data.WarmThroughputs.ElementsAs(ctx, &wts, false)...)
+	if len(wts) > 0 {
+		action.WarmThroughput = expandWarmThroughput(map[string]any{
+			"read_units_per_second":  int(wts[0].ReadUnitsPerSecond.ValueInt64()),
+			"write_units_per_second": int(wts[0].WriteUnitsPerSecond.ValueInt64()),
+		})
+	}
+
+	ut.GlobalSecondaryIndexUpdates = []ddbtypes.GlobalSecondaryIndexUpdate{
+		{
+			Create: action,
+		},
+	}
+
+	response.Diagnostics.Append(validateNewGSIAttributes(data, *table)...)
+
+	if response.Diagnostics.HasError() {
+		return
+	}
+
+	if utRes, err := conn.UpdateTable(ctx, ut); err != nil {
+		response.Diagnostics.AddError(
+			fmt.Sprintf(`Unable to create index "%s" on table "%s"`, data.Name.ValueString(), data.Table.ValueString()),
+			err.Error(),
+		)
+
+		return
+	} else {
+		for _, g := range utRes.TableDescription.GlobalSecondaryIndexes {
+			if aws.ToString(g.IndexName) == data.Name.ValueString() {
+				flattenGSI(ctx, *table, g, &data)
+
+				break
+			}
+		}
+	}
+
+	if _, err = waitTableActive(ctx, conn, data.Table.ValueString(), createTimeout); err != nil {
+		response.Diagnostics.AddError(
+			fmt.Sprintf(`Error while waiting for table "%s" to be active`, data.Table.ValueString()),
+			err.Error(),
+		)
+
+		return
+	}
+
+	if _, err := waitGSIActive(ctx, conn, data.Table.ValueString(), data.Name.ValueString(), r.UpdateTimeout(ctx, data.Timeouts)); err != nil {
+		response.Diagnostics.AddError(
+			fmt.Sprintf(`Error while waiting for GSI "%s" on table "%s" to be active`, data.Name.ValueString(), data.Table.ValueString()),
+			err.Error(),
+		)
+	}
+
+	response.Diagnostics.Append(response.State.Set(ctx, &data)...)
+}
+
+func (r *resourceGlobalSecondaryIndex) Read(ctx context.Context, request resource.ReadRequest, response *resource.ReadResponse) {
+	var data resourceGlobalSecondaryIndexModel
+
+	response.Diagnostics.Append(request.State.Get(ctx, &data)...)
+
+	if response.Diagnostics.HasError() {
+		return
+	}
+
+	re := regexache.MustCompile(":table/([^\\/]+)/index/(.+)")
+	grps := re.FindStringSubmatch(data.ID.ValueString())
+	var tableName string
+	if len(grps) == 3 {
+		tableName = grps[1]
+	} else {
+		tableName = data.Table.ValueString()
+	}
+
+	conn := r.Meta().DynamoDBClient(ctx)
+
+	table, err := findTableByName(ctx, conn, tableName)
+	if err != nil {
+		if tfresource.NotFound(err) {
```

Comment:
`tfresource.NotFound` is deprecated. It should be replaced with `github.com/hashicorp/terraform-provider-aws/internal/retry.NotFound`


```suggestion
		if retry.NotFound(err) {
```